arangod/IResearch/IResearchInvertedIndex.cpp

```diff
@@ -917,16 +887,47 @@ bool IResearchInvertedIndex::matchesDefinition(
 std::unique_ptr<IndexIterator> IResearchInvertedIndex::iteratorForCondition(
     LogicalCollection* collection, transaction::Methods* trx,
     aql::AstNode const* node, aql::Variable const* reference,
-    IndexIteratorOptions const& /*opts*/, int mutableConditionIdx) {
+    IndexIteratorOptions const& opts, int mutableConditionIdx) {
+  if (failQueriesOnOutOfSync() && isOutOfSync()) {
+    THROW_ARANGO_EXCEPTION_MESSAGE(
+        TRI_ERROR_CLUSTER_AQL_COLLECTION_OUT_OF_SYNC,
+        absl::StrCat("link ", std::to_string(id().id()),
+                     " has been marked as failed and needs to be recreated"));
+  }
+
+  auto& state = [trx, this, &opts]() -> IResearchSnapshotState& {
+    auto& state = *(trx->state());
+    // TODO FIXME find a better way to look up a State
+    // we cannot use _index pointer as key - the same is used for storing
+    // removes/inserts so we add 1 to the value (we need just something unique
+    // after all)
+    void const* key = reinterpret_cast<uint8_t const*>(this) + 1;
+    auto* ctx = basics::downCast<IResearchSnapshotState>(state.cookie(key));
+    if (!ctx) {
+      auto ptr = irs::memory::make_unique<IResearchSnapshotState>();
+      ctx = ptr.get();
+      state.cookie(key, std::move(ptr));
+
+      if (opts.waitForSync) {
```

Comment:
If index is used multiple times in the query  - there already might be "snapshot" but if the first usage was without sync - we will skip sync here either. And the worst thing is that here we could not just replace snapshot. If I'm not wrong index executor might not create all iterators before starting the execution, so snapshot might be already in use.